### PR TITLE
Fix the add milestone github action

### PR DESCRIPTION
**Pull request by talldan**
## Description
For some reason the Add Milestone github action stopped working recently. (slack convo - https://wordpress.slack.com/archives/C02QB2JS7/p1623640633051900).

The cause seems to be a change in the error format from the API.

The libraries we're using are quite outdated, so I decided to try updating them. It didn't go completely smoothly.

In particular I found the TypeScript definitions hard to work with, so I'd appreciate some help with that.

## How has this been tested?
Actions should pass

## Types of changes

Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: packages/project-management-automation/lib/tasks/add-milestone/index.js [hunk 1]
```diff
@@ -41,26 +38,19 @@ const isDuplicateValidationError = ( error ) =>
  * @param {string} repo    Repository name.
  * @param {string} title   Milestone title.
  *
- * @return {Promise<OktokitIssuesListMilestonesForRepoResponseItem|void>} Promise resolving to milestone, if exists.
+ * @return {Promise<any|void>} Promise resolving to milestone, if exists.
```
Comment (talldan) [inline packages/project-management-automation/lib/tasks/add-milestone/index.js:41]: Not sure how these types are defined now, but `import('@octokit/rest').IssuesListMilestonesForRepoResponseItem` no longer works.

File: packages/project-management-automation/lib/tasks/add-milestone/index.js [hunk 2]
```diff
@@ -18,20 +17,18 @@ const DAYS_PER_RELEASE = 14;
...
  * @return {boolean} Whether error is a duplicate validation request error.
  */
-const isDuplicateValidationError = ( error ) =>
-	Array.isArray( error.errors ) &&
-	error.errors.some( ( { code } ) => code === 'already_exists' );
+const isDuplicateValidationError = ( requestError ) => {
+	// The included version of RequestError provides no way to access the
+	// full 'errors' array that the github REST API returns. Hopefully they
+	// resolve this soon!
+	return requestError.message.includes( 'already_exists' );
```
Comment (talldan) [inline packages/project-management-automation/lib/tasks/add-milestone/index.js:30]: Unfortunately `@actions/github` uses quite an old version of their `RequestError` library.

**Comment by gwwar** [comment]
Looks like there's existing unit tests, so for folks 👀 we can run this with:

```
npm run test-unit packages/project-management-automation/lib/tasks/add-milestone/test/index.js
```

**Comment by gwwar** [comment]
references for the API endpoint:
https://octokit.github.io/rest.js/v18
https://docs.github.com/en/rest/reference/issues#list-milestones

**Comment by gwwar** [comment]
Example response:
```bash
curl \
>   -H "Accept: application/vnd.github.v3+json" \
>   https://api.github.com/repos/WordPress/gutenberg/milestones
[
  {
    "url": "https://api.github.com/repos/WordPress/gutenberg/milestones/70",
    "html_url": "https://github.com/WordPress/gutenberg/milestone/70",
    "labels_url": "https://api.github.com/repos/WordPress/gutenberg/milestones/70/labels",
    "id": 3734039,
    "node_id": "MDk6TWlsZXN0b25lMzczNDAzOQ==",
    "number": 70,
    "title": "WordPress 5.x",
    "description": "Tasks to be done for a major WordPress release in the near to medium term.",
    "creator": {
      "login": "mtias",
      "id": 548849,
      "node_id": "MDQ6VXNlcjU0ODg0OQ==",
      "avatar_url": "https://avatars.githubusercontent.com/u/548849?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/mtias",
      "html_url": "https://github.com/mtias",
      "followers_url": "https://api.github.com/users/mtias/followers",
      "following_url": "https://api.github.com/users/mtias/following{/other_user}",
      "gists_url": "https://api.github.com/users/mtias/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/mtias/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/mtias/subscriptions",
      "organizations_url": "https://api.github.com/users/mtias/orgs",
      "repos_url": "https://api.github.com/users/mtias/repos",
      "events_url": "https://api.github.com/users/mtias/events{/privacy}",
      "received_events_url": "https://api.github.com/users/mtias/received_events",
      "type": "User",
      "site_admin": false
    },
    "open_issues": 99,
    "closed_issues": 236,
    "state": "open",
    "created_at": "2018-10-12T14:46:17Z",
    "updated_at": "2021-06-07T10:03:10Z",
    "due_on": null,
    "closed_at": null
  },
  {
    "url": "https://api.github.com/repos/WordPress/gutenberg/milestones/150",
    "html_url": "https://github.com/WordPress/gutenberg/milestone/150",
    "labels_url": "https://api.github.com/repos/WordPress/gutenberg/milestones/150/labels",
    "id": 6778712,
    "node_id": "MDk6TWlsZXN0b25lNjc3ODcxMg==",
    "number": 150,
    "title": "Gutenberg 10.8",
    "description": null,
    "creator": {
      "login": "github-actions[bot]",
      "id": 41898282,
      "node_id": "MDM6Qm90NDE4OTgyODI=",
      "avatar_url": "https://avatars.githubusercontent.com/in/15368?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/github-actions%5Bbot%5D",
      "html_url": "https://github.com/apps/github-actions",
      "followers_url": "https://api.github.com/users/github-actions%5Bbot%5D/followers",
      "following_url": "https://api.github.com/users/github-actions%5Bbot%5D/following{/other_user}",
      "gists_url": "https://api.github.com/users/github-actions%5Bbot%5D/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/github-actions%5Bbot%5D/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/github-actions%5Bbot%5D/subscriptions",
      "organizations_url": "https://api.github.com/users/github-actions%5Bbot%5D/orgs",
      "repos_url": "https://api.github.com/users/github-actions%5Bbot%5D/repos",
      "events_url": "https://api.github.com/users/github-actions%5Bbot%5D/events{/privacy}",
      "received_events_url": "https://api.github.com/users/github-actions%5Bbot%5D/received_events",
      "type": "Bot",
      "site_admin": false
    },
    "open_issues": 0,
    "closed_issues": 172,
    "state": "open",
    "created_at": "2021-05-20T20:12:51Z",
    "updated_at": "2021-06-15T14:52:33Z",
    "due_on": "2021-06-02T07:00:00Z",
    "closed_at": null
  },
  {
    "url": "https://api.github.com/repos/WordPress/gutenberg/milestones/151",
    "html_url": "https://github.com/WordPress/gutenberg/milestone/151",
    "labels_url": "https://api.github.com/repos/WordPress/gutenberg/milestones/151/labels",
    "id": 6817780,
    "node_id": "MDk6TWlsZXN0b25lNjgxNzc4MA==",
    "number": 151,
    "title": "Gutenberg 10.9",
    "description": null,
    "creator": {
      "login": "github-actions[bot]",
      "id": 41898282,
      "node_id": "MDM6Qm90NDE4OTgyODI=",
      "avatar_url": "https://avatars.githubusercontent.com/in/15368?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/github-actions%5Bbot%5D",
      "html_url": "https://github.com/apps/github-actions",
      "followers_url": "https://api.github.com/users/github-actions%5Bbot%5D/followers",
      "following_url": "https://api.github.com/users/github-actions%5Bbot%5D/following{/other_user}",
      "gists_url": "https://api.github.com/users/github-actions%5Bbot%5D/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/github-actions%5Bbot%5D/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/github-actions%5Bbot%5D/subscriptions",
      "organizations_url": "https://api.github.com/users/github-actions%5Bbot%5D/orgs",
      "repos_url": "https://api.github.com/users/github-actions%5Bbot%5D/repos",
      "events_url": "https://api.github.com/users/github-actions%5Bbot%5D/events{/privacy}",
      "received_events_url": "https://api.github.com/users/github-actions%5Bbot%5D/received_events",
      "type": "Bot",
      "site_admin": false
    },
    "open_issues": 2,
    "closed_issues": 96,
    "state": "open",
    "created_at": "2021-06-03T01:59:35Z",
    "updated_at": "2021-06-15T14:52:34Z",
    "due_on": "2021-06-16T07:00:00Z",
    "closed_at": null
  }
]
```

**Review by gwwar — APPROVED** [review]
Going to give tentative approval. Let's try to rebase the PR since the performance test and E2E were not green. 

I tested this by making sure units still worked, and checked to see if a real response shape was about right.

🤔 So in theory types should be generated in https://github.com/octokit/openapi-types.ts https://octokit.github.io/types.ts/interfaces/generated_endpoints.endpoints.html#get__repos__owner___repo__milestones

But this doesn't look particularly useful. One thing we could do is define an inline type with what we expect (plus other unknown additional properties). Since this is already broken in trunk, I think it's also fine to use `any` for now and follow up with a better type.

I also wonder if we need to setup a periodic integration test, since units won't catch if GH APIs change.

**Comment by talldan** [comment]
Thanks for the review, will merge and hopefully it'll work 🤞.

**Comment by talldan** [comment]
Awesome, looks like the fix worked. Now to go back and add the milestone to all the PRs that were missed 😄

**Comment by ockham** [comment]
Not sure it helps, but on the note of Octokit type defs, there's been a few related conversations over at https://github.com/WordPress/gutenberg/pull/29043.

**Comment by talldan** [comment]
@ockham Thanks, I didn't realise we had other usages of octokit/rest in the codebase.

I also noticed the `Endpoints` type, but couldn't get that to work for milestones.

**Pull request merged**